[ntuple] add on-disk format blog post (#1053)
root-project/web: blog/_posts/2024-11-28-rntuple-binary-format.md

---
title: "First Release of the RNTuple On-Disk Format"
layout: archive
author: Jakob Blomer
---

You may have heard of RNTuple, from
[CHEP'24](https://indico.cern.ch/event/1338689/contributions/6077632/),
[from](https://indico.jlab.org/event/459/contributions/11594/)
[other](https://indico.cern.ch/event/855454/contributions/4596512/)
[talks](https://indico.cern.ch/event/773049/contributions/3474746/), or from
[our previous ROOT blog post](https://root.cern/blog/rntuple-update/).
In case you haven't: RNTuple is ROOT's new I/O system for event data.
Think of it as TTree, but more compact, faster, modern and more robust.
With respect to TTree, we routinely see file size reductions between 10%-50%, multiple times faster read throughput,
and much better write performance and multicore scalability.
RNTuple can fully harness the performance of modern
[NVMe drives](https://en.wikipedia.org/wiki/NVM_Express) and
[object stores](https://en.wikipedia.org/wiki/Object_storage),
and it comes with a modern, safe, and feature rich API.

After [6 years of R&D](https://indico.cern.ch/event/764726/), RNTuple reached an important milestone:

> we settled on the [first public version of the on-disk binary format](https://github.com/root-project/root/blob/v6-34-00-patches/tree/ntuple/v7/doc/BinaryFormatSpecification.md)!

Reviewer: ```suggestion
we released the [first official version of the on-disk binary format](https://github.com/root-project/root/blob/v6-34-00-patches/tree/ntuple/v7/doc/BinaryFormatSpecification.md)!
```

</center>

This, and other reasons related to the robustness and the use of modern storage systems, convinced us to introduce a new on-disk format after more than 25 years of TTree. With RNTuple, we aim at a new, stable format for the exabytes of data to be expected during the lifetime of upcoming experiments (e.g., the experiments at the HL-LHC, DUNE, and beyond).[[2]](#footnote2)

Obviously, designing for such long time spans requires adding the provisions to adapt to future environments.
For RNTuple, we put mechanisms in place for forward- and backward-compatibility
to enable new software versions to read old data and old software versions to read new data.
For your own data models, you may know such mechanisms under the name "schema evolution".
For the RNTuple format itself, similar techniques are implemented in order to add, for instance,
optional information such as metadata in the future.

Some other key features of the new format are
  - a [format specification](https://github.com/root-project/root/blob/v6-34-00-patches/tree/ntuple/v7/doc/BinaryFormatSpecification.md),
  allowing third parties to easily write data readers and writers without reverse engineering the code;

>   - strict use of 64 bit checksums both on data and metadata;

Reviewer: ```suggestion
  - strict use of 64 bit checksums both on data and metadata, ensuring data integrity;
```

  - forward looking limits that prepare us, e.g.,
  for storing hundreds of thousands of columns in files of hundreds of terabytes each;
  - preparation for native use of object stores such as S3 or DAOS;
  - support for new types, such as `std::variant`, `std::optional`,
  and the upcoming half precision floating point `std::float16_t`.

Now, you may wonder how to actually read and write RNTuples.
To do so, RNTuple comes with [a new set of APIs](https://root.cern.ch/doc/v634/group__tutorial__ntuple.html), too!
Modern, safe, and easy to use correctly (so we hope!).

> The API, in contrast to the on-disk format, is not quite finished yet and did not leave the experimental stage.

Reviewer: ```suggestion
The API, in contrast to the on-disk format, is still evolving thanks to early adopters' feedback and is not quite ready to leave the experimental stage.
```

> An external API review—a first for the ROOT team—conducted by the

Reviewer: ```suggestion
An external API review, _a first for the ROOT team_, conducted by the
```

[High Energy Physics Center for Computational Excellence](https://www.anl.gov/hep-cce) is still just about to conclude.
Expect to see a first set of production APIs in ROOT 6.36, scheduled for the second quarter of 2025.
Meanwhile, please feel free to try out [the current state](https://root.cern/releases/release-63400/)
and tell us what you think!
The ATLAS, CMS, and LHCb experiments already include initial support for the API,
so you may get your next RNTuple from your experiment.
If you have analysis code in RDataFrame: job done.
RDataFrame transparently processes TTree and RNTuple datasets interchangeably.

> To close this post, I'd like to show you the first RNTuple ever written.

Reviewer: Maybe
```suggestion
To close this post, I'd like to show you the first official RNTuple ever written.
```
or
```suggestion
To close this post, I'd like to show you the first RNTuple ever written with the official specification.
```
